tests/queries/0_stateless/02521_tsv_csv_custom_header_detection.sh

```diff
@@ -0,0 +1,305 @@
+
+echo 2
+echo '"x","y","z"
+"UInt32","String","Array(UInt32)"
+123,"Hello","[1,2,3]"
+456,"World","[4,5,6]"' | $CLICKHOUSE_LOCAL --input-format='CSV' --table='test' -q "desc test";
+
+echo '"x","y","z"
+"UInt32","String","Array(UInt32)"
+123,"Hello","[1,2,3]"
+456,"World","[4,5,6]"' | $CLICKHOUSE_LOCAL --input-format='CSV' --table='test' -q "select * from test";
+
+echo 3
+echo '"x","y","z"
+"123","Hello","World"
+"456","World","Hello"' | $CLICKHOUSE_LOCAL --input-format='CSV' --table='test' -q "desc test";
+
+echo '"x","y","z"
+"123","Hello","World"
+"456","World","Hello"' | $CLICKHOUSE_LOCAL --input-format='CSV' --table='test' -q "select * from test";
+
+echo 4
+echo '"x","y","z"
+"UInt32","String","Array(UInt32)"
+"123","Hello","World"
+"456","World","Hello"' | $CLICKHOUSE_LOCAL --input-format='CSV' --table='test' -q "desc test";
+
+echo '"x","y","z"
+"UInt32","String","Array(UInt32)"
+"123","Hello","World"
+"456","World","Hello"' | $CLICKHOUSE_LOCAL --input-format='CSV' --table='test' -q "select * from test";
+
+echo 5
+echo '"x","y","z"' | $CLICKHOUSE_LOCAL --input-format='CSV' --table='test' -q "desc test";
+echo '"x","y","z"'| $CLICKHOUSE_LOCAL --input-format='CSV' --table='test' -q "select * from test";
+
+echo 6
+echo '"x","y","z"
+"UInt32","String","Array(UInt32)"' | $CLICKHOUSE_LOCAL --input-format='CSV' --table='test' -q "desc test";
+
+echo '"x","y","z"
+"UInt32","String","Array(UInt32)"'| $CLICKHOUSE_LOCAL --input-format='CSV' --table='test' -q "select * from test";
+
+echo 7
+echo '"x","y","z"
+"UInt32","String","Array"' | $CLICKHOUSE_LOCAL --input-format='CSV' --table='test' -q "desc test";
+
+echo '"x","y","z"
+"UInt32","String","Array"'| $CLICKHOUSE_LOCAL --input-format='CSV' --table='test' -q "select * from test";
+
+echo 8
+echo '"x","y","z"
+"UInt32","String","Array"
+"123","Hello","[1,2,3]"' | $CLICKHOUSE_LOCAL --input-format='CSV' --table='test' -q "desc test";
+
+echo '"x","y","z"
+"UInt32","String","Array"
+"123","Hello","[1,2,3]"' | $CLICKHOUSE_LOCAL --input-format='CSV' --table='test' -q "select * from test";
+
+echo 9
+echo '"x","y","z"
+123,"Hello","[1,2,3]"' | $CLICKHOUSE_LOCAL --input-format='CSV' --table='test' -q "desc test";
+
+echo '"x","y","z"
+123,"Hello","[1,2,3]"'| $CLICKHOUSE_LOCAL --input-format='CSV' --table='test' -q "select * from test";
+
+echo 10
+echo '"x","y","z"
+123,"""Hello""","[1,2,3]"' | $CLICKHOUSE_LOCAL --input-format='CSV' --table='test' -q "desc test";
+
+echo '"x","y","z"
+123,"""Hello""","[1,2,3]"'| $CLICKHOUSE_LOCAL --input-format='CSV' --table='test' -q "select * from test";
+
+echo 11
+echo '"x","y","z"
+"Hello",\N,"World"' | $CLICKHOUSE_LOCAL --input-format='CSV' --table='test' -q "desc test";
+
+echo '"x","y","z"
+"Hello",\N,"World"'| $CLICKHOUSE_LOCAL --input-format='CSV' --table='test' -q "select * from test";
+
+echo 12
+echo '"x","y","z"
+"Hello",\N,"[1,2,3]"' | $CLICKHOUSE_LOCAL --input-format='CSV' --table='test' -q "desc test";
```

Comment:
Why is this detected as header? `\N` is nullable right? How can we know that first row didn't have NULL by chance

Reply:
Because in the second row we have an array of numbers `[1,2,3]` (in CSV arrays are in double quotes), so we are sure that the first row with all strings is a header